pandas/_libs/lib.pyx

```diff
@@ -2359,3 +2356,103 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
             output[i] = default
 
     return maybe_convert_objects(output)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline void convert_and_set_item(object item, Py_ssize_t index,
+                                      object[:] result,
+                                      bint keep_trivial_numbers):
+    cdef:
+        bint do_convert = 1
+        double double_item
+
+    if keep_trivial_numbers:
+        if isinstance(item, int) and Py_SIZE(item) < 2:
+            if <int>item == 0:
+                do_convert = 0
+        elif isinstance(item, float):
+            double_item = item
+            if double_item == 0.0 or double_item != double_item:
+                do_convert = 0
+
+    if do_convert and not isinstance(item, (str, bytes)):
+        item = PyObject_Str(item)
+
+    result[index] = item
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline void put_object_as_unicode(list lst, Py_ssize_t idx,
+                                       object item):
+    if not isinstance(item, str):
+        item = PyObject_Str(item)
+    lst[idx] = item
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef object _concat_date_cols(tuple date_cols,
+                               bint keep_trivial_numbers=False):
+    cdef:
+        Py_ssize_t i, j, sequence_size = len(date_cols)
+        Py_ssize_t array_size, min_size = 0
+        object[:] result_view
+        flatiter it
```

Comment:
How much mileage does this use of `flatiter` give in terms of perf improvement?  For me at least the flatiter-based code in this module is the most obscure and black-box-ish (and therefore hardest to maintain).  If it can be avoided without defeating the purpose of the PR, I think that would go a long way toward addressing @jreback's comments about added complexity.

Reply:
`flatiter`-based code is about 4x times faster than "normal" iterators for Numpy arrays which Cython generates, as [stated](https://github.com/pandas-dev/pandas/pull/25754#discussion_r268702441) by @anmyachev, and Numpy arrays are the usual input of this function (other sequences' support in this function is actually optional, as AFAIK it's only ever called in currect codebase with Numpy arrays as input).